src/passes/CallArgumentOptimization.cpp

```diff
@@ -0,0 +1,366 @@
+      for (Index i = 0; i < numParams; i++) {
+        Literal value;
+        for (auto* call : calls) {
+          assert(call->target == name);
+          assert(call->operands.size() == numParams);
+          auto* operand = call->operands[i];
+          if (auto* c = operand->dynCast<Const>()) {
+            if (value.type == none) {
+              // This is the first value seen.
+              value = c->value;
+            } else if (value != c->value) {
+              // Not identical, give up
+              value.type = none;
+              break;
+            }
+          } else {
+            // Not a constant, give up
+            value.type = none;
+            break;
+          }
+        }
+        if (value.type != none) {
+          // Success! We can just apply the constant in the function, which makes
+          // the parameter value unused, which lets us remove it later.
+          Builder builder(*module);
+          func->body = builder.makeSequence(
+            builder.makeSetLocal(i, builder.makeConst(value)),
+            func->body
+          );
+          // Mark it as unused, which we know it now is (no point to
+          // re-scan just for that).
+          infoMap[name].unusedParams.insert(i);
+        }
+      }
+    }
+    // Track which functions we changed, and optimize them later if necessary.
+    std::unordered_set<Function*> changed;
+    // We now know which parameters are unused, and can potentially remove them.
+    for (auto& pair : allCalls) {
+      auto name = pair.first;
+      auto& calls = pair.second;
+      auto* func = module->getFunction(name);
+      auto numParams = func->getNumParams();
+      if (numParams == 0) continue;
+      // Iterate downwards, as we may remove more than one.
+      Index i = numParams - 1;
+      while (1) {
+        if (infoMap[name].unusedParams.has(i)) {
+          // Great, it's not used. Check if none of the calls has a param with side
+          // effects, as that would prevent us removing them (flattening before
+          // should have been done).
+          bool can = true;
+          for (auto* call : calls) {
+            auto* operand = call->operands[i];
+            if (EffectAnalyzer(runner->options, operand).hasSideEffects()) {
+              can = false;
+              break;
+            }
+          }
+          if (can) {
+            // Wonderful, nothing stands in our way! Do it.
+            // TODO: parallelize this?
+            removeParameter(func, i, calls);
+            changed.insert(func);
+          }
+        }
+        if (i == 0) break;
+        i--;
+      }
+    }
+    if (optimize && changed.size() > 0) {
+      OptUtils::optimizeAfterInlining(changed, module, runner);
+    }
+  }
+
+private:
+  void removeParameter(Function* func, Index i, std::vector<Call*> calls) {
+    // Clear the type, which is no longer accurate.
+    func->type = Name();
+    // It's cumbersome to adjust local names - TODO don't clear them?
```

Comment:
If we don't clear them here, we'd wind up with things like `(func $foo (param $var$0 i32) (param $var$2 i32))` right? (having deleted `$var$1`)
Is there anything wrong with that? It makes it a bit clearer that this optimization happened, and is less work. Do we depend on names being sequential later on?

Reply:
We'd end up with the wrong names, as removing a parameter decreases the indexes of the ones after it, so the names map would need to be updated for those values.

Could be done, but as this will usually lead to the function being completely reoptimized anyhow, local names will likely be meaningless anyhow (for local coalescing etc.).

Reply:
Makes sense. I thought we used names for everything in binaryen so it would've just worked out, but nope. (Reading more carefully we fix up the *indices* of the get/set local instructions)

Reply:
Yeah, I had initially hoped to use names everywhere, but it just turned out to be too much overhead for locals - so they are numerically indexed.

(But functions and other global things use names everywhere, which I think has worked out well.)